src/ucp/api/ucp.h

```diff
@@ -2479,6 +2593,30 @@ ucs_status_t ucp_worker_set_am_handler(ucp_worker_h worker, uint16_t id,
                                        uint32_t flags);
 
 
+/**
+ * @ingroup UCP_WORKER
+ * @brief Add user defined callback for Active Message.
+ *
+ * This routine installs a user defined callback to handle incoming Active
+ * Messages with a specific id. This callback is called whenever an Active
+ * Message that was sent from the remote peer by @ref ucp_am_send_nbx is
+ * received on this worker.
+ *
+ * @warning Handlers set by this function are not compatible with
+            @ref ucp_am_send_nb routine.
+ *
+ * @param [in]  worker      UCP worker on which to set the Active Message
+ *                          handler.
+ * @param [in]  param       Active Message handler parameters, as defined by
+ *                          @ref ucp_am_handler_param_t.
+ *
+ * @return error code if the worker does not support Active Messages or
+ *         requested callback flags.
+ */
+ucs_status_t ucp_worker_set_am_data_handler(ucp_worker_h worker,
```

Comment:
maybe 'ucp_worker_set_am_recv_handler' ?

Reply:
we don't always call receive. Can get data pointer with zero copy

Reply:
yes, but the callback is handling the receive of active message (either with or w/o separate data)